File: chat/src/main/java/org/thoughtcrime/securesms/jobmanager/impl/NetworkConstraintObserver.java
package org.thoughtcrime.securesms.jobmanager.impl;

import android.app.Application;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;

import androidx.annotation.NonNull;

import org.thoughtcrime.securesms.jobmanager.ConstraintObserver;

public class NetworkConstraintObserver implements ConstraintObserver {

    // 使用字符串常量避免类加载时触发 L.INSTANCE 初始化导致ANR
    private static final String REASON = "NetworkConstraint";

    private final Application application;

    public NetworkConstraintObserver(Application application) {
        this.application = application;
    }

    @Override
    public void register(@NonNull Notifier notifier) {
        application.registerReceiver(new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                NetworkConstraint constraint = new NetworkConstraint.Factory(application).create();

                if (constraint.isMet()) {
                    notifier.onConstraintMet(REASON);
                }
            }
        }, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
    }
}

File: chat/src/main/java/org/thoughtcrime/securesms/jobmanager/impl/NetworkConstraintObserver.kt
package org.thoughtcrime.securesms.jobmanager.impl

import android.app.Application
import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import org.thoughtcrime.securesms.jobmanager.ConstraintObserver

class NetworkConstraintObserver(private val application: Application) : ConstraintObserver {

    override fun register(notifier: ConstraintObserver.Notifier) {
        val cm = application.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val request = NetworkRequest.Builder()
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()
        cm.registerNetworkCallback(request, object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) {
                notifier.onConstraintMet(REASON)
            }
        })
    }

    companion object {
        private const val REASON = "NetworkConstraint"
    }
}
